browser_history/utils.py

```diff
@@ -48,3 +52,107 @@ def get_platform():
     if system == "Windows":
         return Platform.WINDOWS
     raise NotImplementedError(f"Platform {system} is not supported yet")
+
+
+def get_browsers():
+    """This method provides a list of all browsers implemented by
+    browser_history.
+
+    :return: A :py:class:`list` containing implemented browser classes
+        all inheriting from the super class
+        :py:class:`browser_history.generic.Browser`
+
+    :rtype: :py:class:`list`
+    """
+
+    # recursively get all concrete subclasses
+    def get_subclasses(browser):
+        # include browser itself in return list if it is concrete
+        sub_classes = []
+        if not inspect.isabstract(browser):
+            sub_classes.append(browser)
+
+        for sub_class in browser.__subclasses__():
+            sub_classes.extend(get_subclasses(sub_class))
+        return sub_classes
+
+    return get_subclasses(generic.Browser)
+
+
+if get_platform() == Platform.WINDOWS:
+    import winreg  # type: ignore
+
+
+def _default_browser_linux():
+    try:
+        cmd = "xdg-settings get default-web-browser".split()
+        raw_result = subprocess.check_output(cmd, stderr=subprocess.DEVNULL)
+        # most have a suffix ".desktop" so just remove it
+        default = raw_result.decode().strip().lower().replace(".desktop", "")
+    except (FileNotFoundError, subprocess.CalledProcessError, PermissionError):
+        logger.warning("Could not determine default browser")
+        default = None
+
+    return default
+
+
+def _default_browser_win():
+    reg_path = (
+        "Software\\Microsoft\\Windows\\Shell\\Associations\\"
+        "UrlAssociations\\https\\UserChoice"
+    )
+    with winreg.OpenKey(winreg.HKEY_CURRENT_USER, reg_path) as key:
+        default = winreg.QueryValueEx(key, "ProgId")
+        if default is None:
+            logger.warning("Could not determine default browser")
+            return None
+        return default[0].lower()
+
+
+def default_browser():
+    """This method gets the default browser of the current platform
+
+    :return: A :py:class:`browsers.Browser` object representing the default
+        browser in the current platform. If platform is not supported or
+        default browser is unknown or unsupported ``None`` is returned
+
+    :rtype: union[:py:class:`browsers.Browser`, None]
+    """
+    plat = get_platform()
+
+    # ---- get default from specific platform ----
+
+    # Always try to return a lower-cased value for ease of comparison
+    if plat == Platform.LINUX:
+        default = _default_browser_linux()
+    elif plat == Platform.WINDOWS:
+        default = _default_browser_win()
+    else:
+        logger.warning("Default browser feature not supported on this OS")
+        return None
+
+    if default is None:
+        # the current platform has completely failed to provide a default
+        logger.warning("No default browser found")
```

Comment:
There is already a warning from `_default_browser_linux` or `_default_browser_win`, so this warning isn't required I think.

Reply:
Let me remove it asap. But the conditional check is mandatory because if we let the `None` default value through the subsequent steps will try to iterate it 💀. Caught  the issue yesterday thanks to the "`None` return" test you added.

Reply:
Ah right, I meant only the warning.